### PR TITLE
call caching testing.  Closes #12

### LINKED ISSUE #12: Add Call Caching testing to new framework
Enable the following functionality in the new testing framework:

Verifying that cache hits happen as/when expected
Verifying that cache misses happen as/when expected
Verifying the correct value is used when a cache hit happens
Pretty open ended as to how this is done. Modify framework as necessary including input structures, specs, WDLs, etc
AC: Demonstrate that you are successfully testing all of the above conditions in the new framework

### PR DESCRIPTION
**Pull request by ruchim**
Having trouble running the given workflows in this suite through JES, please test on a local backend (for now) with a cromwell where call caching is turned on. Run using sbt test.

### CODE REVIEW

**Comment by ruchim** [comment]
@kshakir @geoffjentry made your proposed changes, so I'm looking for your final say. Thanks!

File: src/main/scala/centaur/CacheFormulas.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    for {
+      s <- TestFormulas.runWorkflowUntilTerminalStatus(request, Succeeded)
+      r <- verifyInputsOutputs(s, request)
+      _ <- verifyCaching(r, request)
+    } yield ()
+  }
+
+  def runSequentialCachingWorkflow(request: WorkflowRequest, secondRequest: WorkflowRequest) = {
+    for {
+      q <- TestFormulas.runWorkflowUntilTerminalStatus(request, Succeeded)
```
Comment (geoffjentry) [inline src/main/scala/centaur/CacheFormulas.scala]: since you're not using `q` it can be `_`

File: src/main/scala/centaur/Metadata.scala [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package centaur
+
+import json.JsonUtils._
+import spray.json._
+import spray.json.{JsObject, DefaultJsonProtocol, JsValue}
+
+object Metadata {
+
+  def makeFilteredMap(metadataJson: String, keySet: Set[String]): Map[String, JsValue] = {
+    metadataJson.parseJson.asJsObject.flatten().fields filter { case (k, v) => keySet exists { x => k.contains(x) } }
```
Comment (geoffjentry) [inline src/main/scala/centaur/Metadata.scala]: since you're not using `v` it can also be an underscore

File: src/main/scala/centaur/Test.scala [hunk 3]
```diff
@@ -117,55 +119,54 @@ object Operations {
...
-    val listOfkeys = Seq("runtimeAttributes", "preemptible", "executionStatus")
-
-    val lastCallAttempt = calls.fields map {
-      case (k1, allCallAttempts: JsArray) =>
-        val lastCallAttempt = allCallAttempts.elements.last.asJsObject
-        val filteredCallAttributes = lastCallAttempt.fields collect {
-          case (k2, callAttributes) if listOfkeys.contains(k2) => k2 -> callAttributes
-        }
-        k1 -> JsObject(filteredCallAttributes)
+  private def makeFilteredMap(metadataJson: String, keySet: Set[String]): Map[String, JsValue] = {
```
Comment (geoffjentry) [inline src/main/scala/centaur/Test.scala]: You added this to the Metadata object above, but left this one here

File: src/main/scala/centaur/TestFormulas.scala [hunk 4]
```diff
@@ -25,6 +25,13 @@ object TestFormulas {
     } yield ()
   }
 
+  def runCachingWorkflow(request: WorkflowRequest) = {
```
Comment (geoffjentry) [inline src/main/scala/centaur/TestFormulas.scala]: Something w/ the same name exists above. Did you mean to remove this?

File: src/test/scala/centaur/CentaurSpec.scala [hunk 5]
```diff
@@ -20,37 +20,37 @@ object FailureTest extends Tag("FailureTest")
...
-    * Test cases which are to be run through runSuccessfulWorkflow.
-    *
-    * Todo: W/ the incoming metadata structure, merge the various failure tests into a single directory
-    */
-  def successfulTestCases(): List[WorkflowRequest] = {
-    val optionalTestCases = CentaurConfig.optionalTestPath map testCases getOrElse List.empty
-    testCases(CentaurConfig.successfulTestCasePath) ++ optionalTestCases
-  }
-}
+//class CentaurSpec extends FlatSpec with Matchers with ParallelTestExecution {
```
Comment (geoffjentry) [inline src/test/scala/centaur/CentaurSpec.scala:23]: I was going to say to uncomment this but it'd actually make my life easier if it remained commented out :)

File: src/main/scala/centaur/Test.scala [hunk 6]
```diff
@@ -180,12 +181,82 @@ object Operations {
...
+    * Verify that none of the calls wihtin the workflow are cached.
+    */
+  def verifyCachingOff (workflow: Workflow, request: WorkflowRequest): Test[Workflow] = {
+    new Test[Workflow] {
+      override def run: Try[Workflow] = {
+        val response = MetadataRequest(Get(CentaurConfig.cromwellUrl + "/api/workflows/v1/" + workflow.id + "/metadata"))
+        sendReceiveFutureCompletion(response  map { allMetadata =>
+          val cacheHitMap = makeFilteredMap(allMetadata, CacheKeys)
+            //allMetadata.parseJson.asJsObject.fields filter { case (k, v) => cacheHit exists { x => k.contains(x) } }
+          if(cacheHitMap.nonEmpty) throw new Exception(s"Found unexpected cache hits for ${request.name}: \n$cacheHitMap")
```
Comment (geoffjentry) [inline src/main/scala/centaur/Test.scala]: space after the `if`

**Comment by geoffjentry** [comment]
@ruchim Does this work on JES now? Just re-noticed your comment from a few days ago. I had some minor moving-stuff-around type of comments but I'm 👍 

[Approved with PullApprove](https://pullapprove.com/broadinstitute/centaur/pull-request/37/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=broadinstitute/centaur)

**Comment by ruchim** [comment]
Hey @geoffjentry -- it still doesn't work on JES since these tests rely on outputs being re-named once they are copied over to the appropriate bucket in JES, and until that bug is fixed we can remove that test and put it in the failing workflows directory?

File: src/main/scala/centaur/CacheFormulas.scala [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+/**
+  * A collection of test formulas which can be used, building upon operations by chaining them together via a
+  * for comprehension. These assembled formulas can then be run by a client
+  */
+object CacheFormulas {
+
+  def runCachingWorkflow(request: WorkflowRequest) = {
+    for {
+      s <- TestFormulas.runWorkflowUntilTerminalStatus(request, Succeeded)
+      r <- verifyInputsOutputs(s, request)
```
Comment (kshakir) [inline src/main/scala/centaur/CacheFormulas.scala]: I'm all for single letter variables like `i` for short iteration loops, usually. I can even do `case (k, v)` for key/value pairs.

But in this case I would request longer variable names specifying what `s` and `r` are abbreviations for.

File: src/main/scala/centaur/CacheFormulas.scala [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    for {
+      q <- TestFormulas.runWorkflowUntilTerminalStatus(request, Succeeded)
+      _ <- runCachingWorkflow(secondRequest)
+    } yield ()
+  }
+
+  def runCachingTurnedOffWorkflow(request: WorkflowRequest) = {
+    for {
+      s <- TestFormulas.runWorkflowUntilTerminalStatus(request, Succeeded)
+      _ <- verifyCachingOff(s, request)
```
Comment (kshakir) [inline src/main/scala/centaur/CacheFormulas.scala]: Same comment on single character variable name.

File: src/main/scala/centaur/json/JsonUtils.scala [hunk 9]
```diff
@@ -50,6 +50,7 @@ object JsonUtils {
     // A couple of helper functions to assist with flattening Cromwell metadata responses
     def hasShardIndex = jsObject.fields.keySet contains "shardIndex"
     def shardIndex = jsObject.fields.get("shardIndex") map { _.toString() }
+    def flattenToMap: Map [String, JsValue] = jsObject.flatten().fields map { case(k, v: JsValue) => k -> v}
```
Comment (kshakir) [inline src/main/scala/centaur/json/JsonUtils.scala]: `case(k` should have a space

**Comment by kshakir** [comment]
👍  after minor updates

[Approved with PullApprove](https://pullapprove.com/broadinstitute/centaur/pull-request/37/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=broadinstitute/centaur)

**Pull request merged**